rust/main/chains/hyperlane-sovereign/src/indexer.rs

```diff
@@ -0,0 +1,131 @@
+use crate::rest_client::{self, Tx, TxEvent};
+use async_trait::async_trait;
+use core::ops::RangeInclusive;
+use hyperlane_core::{
+    ChainCommunicationError, ChainResult, Indexed, Indexer, LogMeta, SequenceAwareIndexer, H256,
+    H512,
+};
+use std::fmt::Debug;
+use tracing::info;
+
+// SovIndexer is a trait that contains default implementations for indexing
+// various different event types on the Sovereign chain to reduce code duplication in
+// e.g. SovereignMailboxIndexer, SovereignInterchainGasPaymasterIndexer, etc.
+#[async_trait]
+pub trait SovIndexer<T>: Indexer<T> + SequenceAwareIndexer<T>
+where
+    T: Into<Indexed<T>> + Debug + Clone + Send,
+{
+    // These are the guys that need to be implemented by the concrete indexer
+    fn client(&self) -> &rest_client::SovereignRestClient;
+    fn decode_event(&self, event: &TxEvent) -> ChainResult<T>;
+    async fn sequence_at_slot(&self, slot: u32) -> ChainResult<Option<u32>>;
+    const EVENT_KEY: &'static str;
+
+    // Default implementation of Indexer<T>
+    async fn fetch_logs_in_range(
+        &self,
+        range: RangeInclusive<u32>,
+    ) -> ChainResult<Vec<(Indexed<T>, LogMeta)>> {
+        info!("Fetching {} logs in range: {:?}", Self::EVENT_KEY, range);
+
+        let mut results = Vec::new();
+
+        for batch_num in range {
+            let batch = self.client().get_batch(u64::from(batch_num)).await?;
+            let batch_hash = parse_hex_to_h256(&batch.hash, "invalid block hash")?;
+            for tx in &batch.txs {
+                let events = self.process_tx(tx, batch_hash)?;
+                results.extend(events);
+            }
```

Comment:
We can avoid nested loops and don't `extend` vector each time by:

```suggestion
            results.extend(
                batch
                    .txs
                    .iter()
                    .flat_map(|tx| self.process_tx(tx, batch_hash))
                    .flatten(),
            );
```